CORE-6173 CORE-6181 Use notary keys for signing/validation

nkovacsx opened the pull request:
### Overview

Currently, we use the ledger key for signing which is wrong. This PR introduces a new `MemberInfo` extension called `notaryKeys`. This property will return the notary keys of the VNode (if it is a notary VNode).

We'll use the **first** element of those notary keys, which will always be the **latest** notary key. This means if there's a key rotation we'll always use the latest key for signing.

When the notary signature is validated we check that the response was signed by one of the notary VNodes. This means, we check that the key is part of the notary composite key.

Test scenario `Notary - Uniqueness client service flow is finishing without exceptions` has been removed since we can only sign on a notary VNode now however, the `test-cordapp` is not installed on the notary so we can't call the `LedgerUniquenessClientService`. This scenario will be tested implicitly by the other scenarios anyway. 

### Test Notes

- Existing smoke test are running successfully as both the signing and validation part has been modified
- Added extra validation for signature in the `NonValidatingNotaryTestFlow` when it is not run through the finality flow
- Removed `UniquenessCheckTestFlow` altogether as this is no longer needed, the uniqueness checker client service will be tested through the plugin
- Added extra test scenario to Utxo finality flow tests - when a signature is received from an identity that is not part of the notary composite key
- Current tests has been modified to use the new API (with the notary keys)
- Added new test scenarios in `LedgerUniquenessCheckerClientServiceImplTest`

nkovacsx committed:
CORE-6173 CORE-6181 Use notary keys for signing/validation

nkovacsx committed:
Merge branch 'release/os/5.0' of https://github.com/corda/corda-runtime-os into nandor/CORE-6173/notary-vnode-key

nkovacsx committed:
detekt

vlajos reviewed (dismissed):
The finality flow/ledger specific parts look OK to me.

relyafi reviewed (commented):
Only one trivial comment to address, others are observations / follow on actions.

I would note however there is a lack of test notes on this PR, beyond saying that a test case has been removed. I'm not really sure why we've removed this - what is the barrier to setting up a proper test scenario using notary virtual nodes if needed?

If possible, I'd also like to see some evidence that the composite key verification logic works when we have multiple notary virtual nodes for the same notary service. The MGM does actively prevent this at present, so may need to be done via unit tests etc.

I'd also add the fact that we've fundamentally changed the key being used to sign, without changing any test cases implies that there are existing gaps in our test coverage.

nkovacsx commented:
@relyafi To answer your questions in order:

1. The `Notary - Uniqueness client service flow is finishing without exceptions` test case was removed because we can only sign in a notary VNode now (as that's the only VNode that has notary keys). 
The only way to execute this test is to call the `UniquenessCheckTestFlow` which is part of the `test-cordapp` CPB which is **NOT** installed on the notary VNode. What I could do is install `test-cordapp` on the notary VNode which is a big overhead because we won't use 90% of the CPB or I could create a separate CPB which **ONLY** includes the `UniquenessCheckTestFlow` which would require a new sub-module. I chose not to do any of these because I feel like this test scenario is only an intermediate step and this scenario will be tested in the other scenarios where the actual plugin is called.

2. `we've fundamentally changed the key being used to sign, without changing any test cases implies that there are existing gaps in our test coverage.` I don't necessarily agree with this. If you check the code, we've changed the key we use for signing, but we also changed the signature verification logic. That means both sides of the process were changed (signing-verifying). Do you think we should add a test scenario that explicitly checks the key that we used for signing?

relyafi commented:
@nkovacsx Ok, understood re. point 1.

Re. point 2 - ok, yes I guess that as long as we are running finality flow, that itself will verify that we have the correct signature. Are we confident that all of the test cases we have that are testing the uniqueness checker client (which is generating the signature) are using finality flow? If so, then I think we're good. If not, those other test cases should probably check that the expected signature is there.

nkovacsx committed:
Add extra test scenario to Finality Flow test

nkovacsx commented:
Thanks @relyafi for your comment, that's actually a good point. Consuming transactions don't go through the finality flow so we should add an extra validation step in our test flow.

nkovacsx committed:
Add extra check to non validating test flow

vlajos reviewed (dismissed):
Ledger/Finality flow parts look good.

relyafi commented on components/ledger/ledger-utxo-flow/src/test/kotlin/net/corda/ledger/utxo/flow/impl/flows/finality/UtxoFinalityFlowTest.kt:
```diff
@@ -69,14 +70,20 @@ class UtxoFinalityFlowTest {
-    private val signatureNotary = digitalSignatureAndMetadata(publicKeyNotary, byteArrayOf(1, 2, 6))
+
+    private val signatureNotary = digitalSignatureAndMetadata(publicKeyNotaryVNode1, byteArrayOf(1, 2, 6))
+    private val invalidNotarySignature = digitalSignatureAndMetadata(invalidNotaryVNodeKey, byteArrayOf(1, 2, 3))
```
I'm assuming the byte array should be different from `signatureAlice1`, just to be safe?

nkovacsx replied:
Good point!

nkovacsx committed:
Change signature

relyafi reviewed (dismissed):
LGTM

ronanbrowne reviewed (dismissed):
lgtm - build

nkovacsx committed:
Merge branch 'release/os/5.0' of https://github.com/corda/corda-runtime-os into nandor/CORE-6173/notary-vnode-key

 Conflicts:
	components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/finality/UtxoFinalityBase.kt

nkovacsx committed:
Pass notary service keys to uniqueness client service

vlajos reviewed (dismissed):
Ledger/Finality flow parts reapproved.

relyafi commented on components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/finality/UtxoFinalityBase.kt:
```diff
@@ -45,6 +45,7 @@ abstract class UtxoFinalityBase : SubFlow<UtxoSignedTransaction> {
         onFailure: ((message: String) -> Unit)? = null
     ) {
         try {
+            log.debug("Verifying signature($signature) of transaction: $transactionId")
```
```suggestion
            log.debug { "Verifying signature($signature) of transaction: $transactionId" }
```

relyafi replied:
More efficient as will only be evaluated when debug logging is enabled.

relyafi commented on notary-plugins/notary-plugin-non-validating/notary-plugin-non-validating-server/src/main/kotlin/com/r3/corda/notary/plugin/nonvalidating/server/NonValidatingNotaryServerFlowImpl.kt:
```diff
@@ -101,13 +108,21 @@ class NonValidatingNotaryServerFlowImpl() : ResponderFlow {
 
             verifyTransaction(requestPayload)
 
+            val notaryServiceKey = notaryLookup.notaryServices.first().publicKey
+
+            // If the key is a composite key (multiple notary VNodes) we need to extract the leaves
+            // If the key is not a composite key (single notary VNode) we can simply use that public key
+            val notaryServicePublicKeys = (notaryServiceKey as? CompositeKey)?.leafKeys?.toList()
+                ?: listOf(notaryServiceKey)
```
This looks incorrect - we can't assume only one notary service on the network.

I believe @mnesbit 's intent (which solves this problem) is that the client node sends the notary service key across as part of the payload from client to server. Then the server plugin has no selection to do - it's provided with the service key the client is expecting a signature from.

nkovacsx replied:
hmm, okay, that makes sense

relyafi reviewed (changes requested):
I think there needs to be additional logic to send the notary service key from client to server flows.

Also, are we missing test cases to cover this new behavior, specifically:

- Testing that we can successfully notarise when the composite key contains multiple leaf keys (since we can't currently test this in a real system). This test should also should also ensure that the notary vnode used to notarise is not the first in that list of leaf keys (since incorrectly using `first()` is a common bug)
- Testing that rejection occurs if the client specifies a notary service key with leaf keys that the notary vnode does not have a corresponding private key for (which simulates routing to the wrong vnode for a notary service)

nkovacsx committed:
Add notary key to payload, new tests

relyafi commented on components/uniqueness/uniqueness-checker-client-service-impl/src/test/kotlin/net/corda/uniqueness/client/impl/LedgerUniquenessCheckerClientServiceImplTest.kt:
```diff
@@ -0,0 +1,224 @@
+    private val argumentCaptor = argumentCaptor<Class<out UniquenessCheckExternalEventFactory>>()
+
+    @Test
+    fun `Uniqueness check client responds with success if uniqueness check was successful`() {
```
This test case appears to be redundant, as the next test case is identical?

relyafi commented on components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/finality/UtxoFinalityBase.kt:
```diff
@@ -45,7 +45,9 @@ abstract class UtxoFinalityBase : SubFlow<UtxoSignedTransaction> {
     ) {
         try {
-            log.debug("Verifying signature($signature) of transaction: $transactionId")
+            log.debug {
+                "Verifying signature($signature) of transaction: $transactionId"
+            }
```
```suggestion
            log.debug { "Verifying signature($signature) of transaction: $transactionId" }
```

Consistent formatting with debug statement 2 lines further down.

relyafi reviewed (commented):
Couple of minor comments, then LGTM.

@mnesbit I think this change should now address most of your concerns - the only place that we're still using member info is to lookup the key on the notary vnode, which I think is unavoidable until the crypto API is updated.

nkovacsx committed:
minor fixes

nkovacsx committed:
fix test

relyafi reviewed (dismissed):
LGTM

nkovacsx committed:
bump version

nkovacsx committed:
Merge branch 'release/os/5.0' of https://github.com/corda/corda-runtime-os into nandor/CORE-6173/notary-vnode-key

 Conflicts:
	gradle.properties

ronanbrowne reviewed (approved):
lgtm - build
